File: src/main/java/komu/demodel/domain/ClassModule.java
/*
 * Copyright (C) 2010 Juha Komulainen. All rights reserved.
 */
package komu.demodel.domain;

import static java.util.Collections.emptyList;

import java.util.List;

public final class ClassModule extends Module {

    private boolean programModule = false;
    
    /** Detailed information about the type of this module, or null if we haven't visited it. */
    private TypeType type;
    
    public ClassModule(String name, PackageModule parent) {
        super(name, parent);
    }
   
    public void markAsProgramModule(TypeType type) {
        programModule = true;
        this.type = type;
    }
    
    @Override
    PackageModule getRoot() {
        return parent.getRoot();
    }

    public boolean isAbstract() {
        return type != null && type.isAbstract();
    }
    
    @Override
    public ModuleType getType() {
        return ModuleType.TYPE;
    }
    
    @Override
    public boolean isProgramModule() {
        return programModule;
    }
    
    @Override
    public boolean isLeaf() {
        return true;
    }
    
    @Override
    public int countLeafsUnder() {
        return 1;
    }
    
    @Override
    public void filterNonProgramReferences() {
    }
    
    @Override
    public List<Module> getChildren() {
        return emptyList();
    }
    
    @Override
    public void sortChildren() {
    }
}

File: src/main/scala/komu/demodel/domain/ClassModule.scala
/*
 * Copyright (C) 2010 Juha Komulainen. All rights reserved.
 */
package komu.demodel.domain

final class ClassModule(name: String, _parent: PackageModule) extends Module(name, Some(_parent)) {

  /** Detailed information about the type of this module, or null if we haven't visited it. */
  private var kind: Option[TypeType] = None

  def markAsProgramModule(kind: TypeType) {
    this.kind = Some(kind)
  }

  def root = _parent.root
  def isAbstract = kind.map(_.isAbstract).getOrElse(false)
  def moduleType = ModuleType.TYPE  
  def isProgramModule = kind.isDefined
  def isLeaf = true
  def countLeafsUnder = 1
  def filterNonProgramReferences() { }
  def children = List()
  def sortChildren() { }
}
